addons/account/models/account_account.py

```diff
@@ -936,56 +950,72 @@ def name_create(self, name):
-                    check_company_vals['company_ids'] = [Command.link(company.id) for company in companies[1:]]
-                if check_company_vals:
-                    new_account.write(check_company_vals)
+                if vals.get('code_mapping_ids') is False:
+                    vals['code_mapping_ids'] = []
+
+                if vals.get('code_mapping_ids'):
+                    # Prevent anything other than CREATE in `code_mapping_ids`
+                    vals['code_mapping_ids'] = [command for command in (vals['code_mapping_ids'] or []) if command[0] == Command.CREATE]
+                    code_by_company = {command[2].get('company_id'): command[2].get('code') for command in vals['code_mapping_ids']}
+
+                    # We can only have values either in `code` or in `code_mapping_ids`, otherwise `code` will get
+                    # written first and will then be protected, which will interfere with `code_mapping_ids.create`.
+                    # Therefore, if we have values in both, we put the value from `code` into `code_mapping_ids`.
+                    if 'code' in vals:
+                        code_in_active_company = vals.pop('code')
+                        if companies[0].id not in code_by_company:
+                            vals['code_mapping_ids'].append(Command.create({'company_id': companies[0].id, 'code': code_in_active_company}))
+                            code_by_company[companies[0].id] = code_in_active_company
+
+                    # Since we're giving values for `code_mapping_ids` and not `code`, `_compute_account_type` won't be called before
+                    # the record is created in DB, so the NOT NULL constraint on `account_type` will get triggered.
+                    # We therefore look here for a closest parent account in any of the companies to which the new account will belong,
+                    # and assign an appropriate account type.
+                    if 'account_type' not in vals:
+                        for company in companies:
+                            if code_in_company := code_by_company.get(company):
+                                company_accounts = accounts_by_company[company].with_company(company)
+                                if closest_parent_account := self._get_closest_parent_account(code_in_company, company_accounts):
+                                    vals['account_type'] = closest_parent_account.account_type
+                                    break
+                        else:
+                            vals['account_type'] = 'asset_current'
+
+            new_accounts = super(AccountAccount, self.with_context({
+                **self.env.context,
+                'allowed_company_ids': companies.ids,
+                'defer_account_code_checks': True,
+                # Don't get a default value for `code_mapping_ids` from default_get
+                'default_code_mapping_ids': self.env.context.get('default_code_mapping_ids', []),
```

Comment:
What would the default be? Why is it an issue?

Reply:
The default is to create code mappings with empty codes. It's used to initialize the One2many for a new account.

If I could redesign this whole 'virtual record' though, I'd redefine `code_mapping_ids` as a computed Many2many, create it via a compute in `account.account`, catch commands on `code_mapping_ids` directly in the create/write methods on `account.account` and then we wouldn't need the `_search` override anymore and we wouldn't need to create records in cache. But I fear changing the One2many to a computed Many2many would break stable policy, what do you think?